Firestore/core/src/firebase/firestore/local/leveldb_mutation_queue.mm

```diff
@@ -0,0 +1,519 @@
+  // it.
+  check_iterator->Seek(key);
+  HARD_ASSERT(check_iterator->Valid(), "Mutation batch %s did not exist",
+              DescribeKey(key));
+
+  HARD_ASSERT(key == check_iterator->key(),
+              "Mutation batch %s not found; found %s", DescribeKey(key),
+              DescribeKey(check_iterator->key()));
+
+  db_.currentTransaction->Delete(key);
+
+  for (FSTMutation* mutation in batch.mutations) {
+    key = LevelDbDocumentMutationKey::Key(user_id_, mutation.key, batch_id);
+    db_.currentTransaction->Delete(key);
+    [db_.referenceDelegate removeMutationReference:mutation.key];
+  }
+}
+
+const std::vector<FSTMutationBatch*>
+LevelDbMutationQueue::AllMutationBatches() {
+  std::string user_key = LevelDbMutationKey::KeyPrefix(user_id_);
+
+  auto it = db_.currentTransaction->NewIterator();
+  it->Seek(user_key);
+  std::vector<FSTMutationBatch*> result;
+  for (; it->Valid() && absl::StartsWith(it->key(), user_key); it->Next()) {
+    result.push_back(ParseMutationBatch(it->value()));
+  }
+  return result;
+}
+
+std::vector<FSTMutationBatch*>
+LevelDbMutationQueue::AllMutationBatchesAffectingDocumentKeys(
+    const DocumentKeySet& document_keys) {
+  // Take a pass through the document keys and collect the set of unique
+  // mutation batch_ids that affect them all. Some batches can affect more than
+  // one key.
+  std::set<BatchId> batch_ids;
+
+  auto index_iterator = db_.currentTransaction->NewIterator();
+  LevelDbDocumentMutationKey row_key;
+  for (const DocumentKey& document_key : document_keys) {
+    std::string index_prefix =
+        LevelDbDocumentMutationKey::KeyPrefix(user_id_, document_key.path());
+    for (index_iterator->Seek(index_prefix); index_iterator->Valid();
+         index_iterator->Next()) {
+      // Only consider rows matching exactly the specific key of interest. Index
+      // rows have this form (with markers in brackets):
+      //
+      // <User>user <Path>collection <Path>doc <BatchId>2 <Terminator>
+      // <User>user <Path>collection <Path>doc <BatchId>3 <Terminator>
+      // <User>user <Path>collection <Path>doc <Path>sub <Path>doc <BatchId>3
+      // <Terminator>
+      //
+      // Note that Path markers sort after BatchId markers so this means that
+      // when searching for collection/doc, all the entries for it will be
+      // contiguous in the table, allowing a break after any mismatch.
+      if (!absl::StartsWith(index_iterator->key(), index_prefix) ||
+          !row_key.Decode(index_iterator->key()) ||
+          row_key.document_key() != document_key) {
+        break;
+      }
+
+      batch_ids.insert(row_key.batch_id());
+    }
+  }
+
+  return AllMutationBatchesWithIds(batch_ids);
+}
+
+std::vector<FSTMutationBatch*>
+LevelDbMutationQueue::AllMutationBatchesAffectingDocumentKey(
+    const DocumentKey& key) {
+  // Scan the document-mutation index starting with a prefix starting with the
```

Comment:
There's a lot of duplication between this function and the previous one.

Could this function be reimplemented like this?

```
DocumentKeySet keys{key}
return AllMutationBatchesAffectingDocumentKeys(keys);
```

Reply:
Yeah, good catch. It's two separate methods in the ObjC source, but I'm not sure why. I've applied this refactor.